File: packages/contracts/python/noa_contracts/render_scene.py
from __future__ import annotations

from typing import Literal, Any
from pydantic import BaseModel


class Camera(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class RenderScene(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    id: str
    canonical_model_id: str
    style_profile_id: str | None
    cameras: list[Camera]
    lighting_setup: dict[str, Any]
    environment: dict[str, Any]
    material_assignments: list[dict[str, Any]]
    render_settings: dict[str, Any]  # engine, samples, resolution, output_formats
    variants: list[dict[str, Any]]
    created_at: str

File: packages/contracts/v1/render-scene.ts
import type { UUID, ISO8601 } from "./common.js";

export interface Camera {
  id: string;
  [key: string]: unknown;
}

export interface LightingSetup {
  [key: string]: unknown;
}

export interface EnvironmentMap {
  [key: string]: unknown;
}

export interface MaterialAssignment {
  [key: string]: unknown;
}

export interface RenderVariant {
  id: string;
  [key: string]: unknown;
}

export interface RenderScene {
  schema_version: "1.0";
  id: UUID;
  canonical_model_id: UUID;
  style_profile_id: UUID | null;
  cameras: Camera[];
  lighting_setup: LightingSetup;
  environment: EnvironmentMap;
  material_assignments: MaterialAssignment[];
  render_settings: {
    engine: "cycles" | "eevee";
    samples: number;
    resolution: { width: number; height: number };
    output_formats: ("jpeg" | "png")[];
  };
  variants: RenderVariant[];
  created_at: ISO8601;
}
